src/qiskit_qasm3_import/state.py

```diff
@@ -0,0 +1,205 @@
+}
+
+
+class AddressingMode:
+    """Addressing mode for qubits in OpenQASM 3 programs.
+
+    This class is useful as long as we allow only physical or virtual addressing modes, but not
+    mixed. If the latter is supported in the future, this class will be modified or removed.
+
+    """
+
+    _Mode = enum.Enum("_Mode", ["UNKNOWN", "PHYSICAL", "VIRTUAL"])
+    (UNKNOWN, PHYSICAL, VIRTUAL) = (_Mode.UNKNOWN, _Mode.PHYSICAL, _Mode.VIRTUAL)
+
+    def __init__(self):
+        self._state = self.UNKNOWN
+
+    def set_physical_mode(self, node):
+        """Set the addressing mode to physical. On success return `True`, otherwise raise an
+        exception."""
+        if self._state == self.VIRTUAL:
+            raise_from_node(
+                node,
+                "Physical qubit referenced in virtual addressing mode. Mixing "
+                "modes not currently supported.",
+            )
+        self._state = self.PHYSICAL
+
+    def set_virtual_mode(self, node):
+        """Set the addressing mode to virtual.
+        On success return `True`, otherwise raise an exception."""
+        if self._state == self.PHYSICAL:
+            raise_from_node(
+                node,
+                "Virtual qubit declared in physical addressing mode. Mixing modes not currently "
+                "supported.",
+            )
+        self._state = self.VIRTUAL
+
+
+class SymbolTable:
+    __slots__ = ("symbols", "scope", "base", "global_symbols")
+
+    def __init__(self, scope: Scope, base: Optional["SymbolTable"] = None):
+        self.symbols = {}
+        self.scope = scope
+        self.base = base
+        self.global_symbols = base.global_symbols if base is not None else _BUILTINS.copy()
+
+    def __contains__(self, name: str):
+        return name in (self.symbols, self.global_symbols) or (
+            self.base is not None and name in self.base.symbols
+        )
+
+    def get(self, name: str, node=None):
+        """Lookup symbol `name`
+
+        Return `None` if no symbol is found. Raise an exception if the symbol is
+        inaccessible in the scope of this table.
+        """
+        if (symbol := self.symbols.get(name, None)) is not None:
+            return symbol
+        if self.base is not None:
+            if (symbol := self.base.get(name, None)) is None:
+                return None
+            return self._check_visible(symbol, node)
+        return self.global_symbols.get(name, None)
+
+    def _check_visible(self, symbol, node):
+        if (
+            symbol.scope is Scope.BUILTIN
+            or self.scope is not Scope.GATE
+            or symbol.scope is not Scope.GLOBAL
+        ):
+            return symbol
+        if self.scope is Scope.GATE:
+            if (
+                isinstance(symbol.type, types.Gate)
+                or isinstance(
+                    symbol.type, (types.Int, types.Uint, types.Float, types.Angle, types.Duration)
+                )
+                and symbol.type.const
```

Comment:
Super minor: can we parenthesis the `isinstance(symbol.type, (...)) and symbol.type.const` check?  I know it's valid by the operator precedence once I think about it, but personally it's not something that I can just "see" automatically in the same way as mathematical rules.

Reply:
I agree. I copied it from here ( :laughing-and-crying-while-rolling-on-the-floor: )     
https://github.com/Qiskit/qiskit-qasm3-import/blob/246e999573c0d7a2162b779d9525b8b5b79c3cf0/src/qiskit_qasm3_import/converter.py#L107-L116

But seriously, it's not so minor. I think it will make it easier for the next person to read.

Reply:
Fixed in e894ba1

Reply:
Ha, guess it was me who did it originally then!  Thanks for the change.